admin/static/js/sales_analytics/dashboard.js

```diff
@@ -0,0 +1,430 @@
+                                    '#00d1c1',
+                                    'Users');
+    };
+
+    self.getRepeatActionCountMonthly = function(repeatActionUserMonthly) {
+        var chart = self.drawMetric('db-chart-repeat-action-monthly',
+                                    repeatActionUserMonthly.repeat_action_count,
+                                    '#FFD37F',
+                                    'Users');
+    };
+
+    self.getTotalUserCount = function(userCount) {
+        var chart = self.drawMetric('db-chart-total-user',
+                                    userCount.total,
+                                    '#005c66',
+                                    'Users');
+    };
+
+    self.prepareChart = function(elementId) {
+        var chart = new keen.Dataviz();
+        return chart.el(document.getElementById(elementId)).prepare();
+    };
+
+    self.drawChart = function(chart, type, title, result, color) {
+        chart.attributes({title: title, width: '100%'});
+        chart.adapter({chartType: type});
+        chart.parseRawData({result: result});
+        chart.render();
+    };
+
+    self.drawMetric = function(elementId, result, color, title) {
+        var chart = new keen.Dataviz()
+          .el(document.getElementById(elementId))
+          .parseRawData({ result: result })
+          .chartType('metric')
+          .colors([color])
+          .title(title)
+          .render();
+    };
+
+    self.init = function() {
+        console.log('init');
+        keen.ready(self.run());
+    };
+
+    self.run = function() {
+        console.log('run');
+        self.getOSFProductUsage();
+        self.getAverageUserSessionLength();
+        self.getAverageMAUSessionLength();
+        self.getAverageUserSessionHistory(true, 12, null, null, null);
+        self.getUserCount(userCount);
+        self.getUserPercentage(userCount);
+        self.getMultiProductCountYearly(multiProductMetricsYearly);
+        self.getMultiProductCountMonthly(multiProductMetricsMonthly);
+        self.getCrossProductCountYearly(multiProductMetricsYearly);
+        self.getCrossProductCountMonthly(multiProductMetricsMonthly);
+        self.getMultiActionCountYearly(multiProductMetricsYearly);
+        self.getMultiActionCountMonthly(multiProductMetricsMonthly);
+        self.getRepeatActionCountMonthly(repeatActionUserMonthly);
+        self.getTotalUserCount(userCount);
+    };
+
+    self.clean = function() {
+        console.log('clean');
+    };
+
+    self.extractDataSet = function(keenResult) {
+        if (!keenResult) {
+            return 0;
+        }
+
+        var beginTime;
+        var endTime;
+        var deltaSet = [];
+
+        for (var i in keenResult.result) {
+            var session = keenResult.result[i];
+            if (session.hasOwnProperty('result')) {
+                if (session.result.length === 1) {
+                    // TODO: take care of the situation where there is only one 'keen.timestamp'
```

Comment:
I was half wrong. This single timestamp issue can happen a lot when we only track page refresh / new page event. Gage is not adding any new tracking based on user actions right now. So I still need to fix this part. :)